File: grpc_server.py
import grpc
from concurrent import futures

from google.protobuf.empty_pb2 import Empty

import music_pb2
import music_pb2_grpc

from database import SessionLocal
from models.streaming_model import User, Track, Playlist


class MusicService(music_pb2_grpc.MusicServiceServicer):

    # ==========================
    #         USERS
    # ==========================
    def ListUsers(self, request, context):
        db = SessionLocal()
        users = db.query(User).all()

        response = music_pb2.UserList()

        for u in users:
            user = response.users.add()
            user.id = u.id
            user.username = u.username
            user.full_name = u.full_name or ""
            user.age = getattr(u, "age", 0)

        return response

    # ==========================
    #         TRACKS
    # ==========================
    def ListTracks(self, request, context):
        db = SessionLocal()
        tracks = db.query(Track).all()

        response = music_pb2.TrackList()

        for t in tracks:
            track = response.tracks.add()
            track.id = t.id
            track.title = t.title
            track.artist = t.artist or ""
            track.file_path = t.file_path or ""

        return response

    # ==========================
    #     PLAYLISTS OF USER
    # ==========================
    def PlaylistsOfUser(self, request, context):
        db = SessionLocal()
        user = db.query(User).filter(User.id == request.id).first()

        if not user:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("User not found")
            return music_pb2.PlaylistList()

        response = music_pb2.PlaylistList()

        for p in user.playlists:
            playlist = response.playlists.add()
            playlist.id = p.id
            playlist.name = p.name
            playlist.owner_id = p.owner_id

        return response

    # ==========================
    #     TRACKS OF PLAYLIST
    # ==========================
    def TracksOfPlaylist(self, request, context):
        db = SessionLocal()
        playlist = db.query(Playlist).filter(Playlist.id == request.id).first()

        if not playlist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("Playlist not found")
            return music_pb2.TrackList()

        response = music_pb2.TrackList()

        for t in playlist.tracks:
            track = response.tracks.add()
            track.id = t.id
            track.title = t.title
            track.artist = t.artist or ""
            track.file_path = t.file_path or ""

        return response

    # ==========================
    #   PLAYLISTS CONTAINING TRACK
    # ==========================
    def PlaylistsContainingTrack(self, request, context):
        db = SessionLocal()
        track = db.query(Track).filter(Track.id == request.id).first()

        if not track:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("Track not found")
            return music_pb2.PlaylistList()

        response = music_pb2.PlaylistList()

        for p in track.playlists:
            playlist = response.playlists.add()
            playlist.id = p.id
            playlist.name = p.name
            playlist.owner_id = p.owner_id

        return response

    # ==========================
    #         TRACK INFO
    # ==========================
    def TrackInfo(self, request, context):
        db = SessionLocal()
        t = db.query(Track).filter(Track.id == request.id).first()

        if not t:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("Track not found")
            return music_pb2.Track()

        return music_pb2.Track(
            id=t.id,
            title=t.title,
            artist=t.artist or "",
            file_path=t.file_path or ""
        )


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    music_pb2_grpc.add_MusicServiceServicer_to_server(MusicService(), server)
    server.add_insecure_port("[::]:50051")
    print("gRPC server running at 0.0.0.0:50051")
    server.start()
    server.wait_for_termination()


if __name__ == "__main__":
    serve()

File: grpc_server.js
import grpc from "@grpc/grpc-js";
import protoLoader from "@grpc/proto-loader";
import { Sequelize, DataTypes } from "sequelize";

const PROTO_PATH = "./streaming.proto";

// ==============================================
//   CARREGA O PROTO
// ==============================================

const pkgDef = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});

const proto = grpc.loadPackageDefinition(pkgDef).music;

// ==============================================
//   BANCO DE DADOS
// ==============================================

const sequelize = new Sequelize({
  dialect: "sqlite",
  storage: "./music_service.db",
  logging: false,
});

const User = sequelize.define(
  "User",
  {
    username: DataTypes.STRING,
    full_name: DataTypes.STRING,
    age: DataTypes.INTEGER,
  },
  { timestamps: false }
);

const Track = sequelize.define(
  "Track",
  {
    title: DataTypes.STRING,
    artist: DataTypes.STRING,
    file_path: DataTypes.STRING,
  },
  { timestamps: false }
);

const Playlist = sequelize.define(
  "Playlist",
  {
    name: DataTypes.STRING,
    owner_id: DataTypes.INTEGER, // certifique-se de que essa coluna exista no DB
  },
  { timestamps: false }
);

// ==============================================
//   RELACIONAMENTOS
// ==============================================

// Usuário → Playlist
User.hasMany(Playlist, { foreignKey: "owner_id" });
Playlist.belongsTo(User, { foreignKey: "owner_id" });

// Playlist ↔ Track
Playlist.belongsToMany(Track, {
  through: "playlist_tracks",
  foreignKey: "playlist_id",
  otherKey: "track_id",
  timestamps: false, // desativa createdAt/updatedAt
});

Track.belongsToMany(Playlist, {
  through: "playlist_tracks",
  foreignKey: "track_id",
  otherKey: "playlist_id",
  timestamps: false,
});

// ==============================================
//   IMPLEMENTAÇÃO DOS ENDPOINTS
// ==============================================

const service = {
  async GetUsers(_, callback) {
    const users = await User.findAll();
    callback(null, { users });
  },

  async GetTracks(_, callback) {
    const tracks = await Track.findAll();
    callback(null, { tracks });
  },

  async PlaylistsOfUser({ user_id }, callback) {
    const whereClause = {};
    if (user_id != null) {
      whereClause.owner_id = user_id;
    }

    const playlists = await Playlist.findAll({
      where: whereClause,
      include: [
        { model: User, attributes: ["id", "username", "full_name", "age"] },
        { model: Track, attributes: ["id", "title", "artist", "file_path"] },
      ],
    });

    callback(null, { playlists });
  },

  async TracksOfPlaylist({ playlist_id }, callback) {
    const playlist = await Playlist.findByPk(playlist_id);
    const tracks = playlist
      ? await playlist.getTracks({
          attributes: ["id", "title", "artist", "file_path"],
        })
      : [];
    callback(null, { tracks });
  },

  async PlaylistsContainingTrack({ track_id }, callback) {
    const track = await Track.findByPk(track_id);
    const playlists = track
      ? await track.getPlaylists({
          include: [
            { model: User, attributes: ["id", "username", "full_name"] },
          ],
        })
      : [];
    callback(null, { playlists });
  },

  async TrackInfo({ track_id }, callback) {
    const track = await Track.findByPk(track_id);
    callback(null, { track });
  },
};

// ==============================================
//   SERVIDOR gRPC
// ==============================================

function start() {
  const server = new grpc.Server();

  server.addService(proto.MusicService.service, service);
  server.bindAsync(
    "localhost:50052",
    grpc.ServerCredentials.createInsecure(),
    () => {
      server.start();
      console.log("🔥 Servidor gRPC JS rodando em: localhost:50052");
    }
  );
}

start();
